### PR TITLE
Add missing Screen.Equals/GetHashCode overrides

### PR DESCRIPTION
**Pull request by MrJul**
This PR adds missing overrides for `Equals(object)` and `GetHashCode()` on `Screen`.
This has two purposes: first to fix some build warnings, and second to correct the screen equality behavior.

Before this PR, `Screen.Equals(Screen)` was delegating to `base.Equals(object)`. I think it was meant to call the `Equals(object)` overridden in the derived `PlatformScreen`, but the `base` call make it _non_-virtual. 

Basically, `Equals(Screen)` was always equivalent to `ReferenceEquals()` and thus different from `Equals(object)`.

`Equals(object)` now calls `Equals(Screen)`, which is overridden.
Ideally, we should make `Screen`, `Screen.Equals(Screen)` and `Screen.GetHashCode()` abstract in v12.

### CODE REVIEW

**Comment by maxkatz6** [comment]
> This PR adds missing overrides for Equals(object) and GetHashCode() on Screen.

[These were removed from my refactoring PR](https://github.com/AvaloniaUI/Avalonia/pull/16295#discussion_r1676800885), as implementation was relying on mutable properties, when IPlatformHandle wasn't available. Which was the case with Linux initially.
Now, since Linux backend provides a proper IPlatformHandle, these overloads make more sense.

> Ideally, we should make Screen, Screen.Equals(Screen) and Screen.GetHashCode() abstract in v12.

Ideally, I would make Screen class sealed (or with an internal ctor). And only (i.e. - **only**) created by the windowing backend, always backed by the IPlatformHandle.

**Review by maxkatz6 — APPROVED** [review]
